File: go/1883.go
package main

import (
	"math"
)

func minSkips(dist []int, speed int, hoursBefore int) int {
	const EPS = 1e-7
	n := len(dist)
	dp := make([][]float64, n+1)
	for i := range dp {
		dp[i] = make([]float64, n+1)
		for j := range dp[i] {
			dp[i][j] = math.Inf(1)
		}
	}
	dp[0][0] = 0.0
	for i := 1; i <= n; i++ {
		for j := 0; j <= i; j++ {
			if j != i {
				dp[i][j] = min(dp[i][j], math.Ceil(dp[i-1][j]+float64(dist[i-1])/float64(speed)-EPS))
			}
			if j != 0 {
				dp[i][j] = min(dp[i][j], dp[i-1][j-1]+float64(dist[i-1])/float64(speed))
			}
		}
	}
	for i := 0; i < n+1; i++ {
		if dp[n][i] < float64(hoursBefore)+EPS {
			return i
		}
	}
	return -1
}

File: python/1883.py
from math import ceil
from typing import List

class Solution:
    def minSkips(self, dist: List[int], speed: int, hoursBefore: int) -> int:
        EPS = 1e-7
        n = len(dist)
        dp = [[float("inf")] * (n + 1) for _ in range(n + 1)]
        dp[0][0] = 0.
        for i in range(1, n+1):
            for j in range(i+1):
                if j != i:
                    dp[i][j] = min(dp[i][j], ceil(dp[i-1][j] + dist[i-1] / speed - EPS))
                if j != 0:
                    dp[i][j] = min(dp[i][j], dp[i-1][j-1] + dist[i-1] / speed)
        for j in range(n + 1):
            if dp[n][j] < hoursBefore + EPS:
                return j
        return -1